--- winrt/Windows.System.h
+++ patched/Windows.System.h
@@ -422,6 +422,7 @@
 
 WINRT_EXPORT namespace winrt
 {
+#if 0
     [[nodiscard]] inline auto resume_foreground(
         Windows::System::DispatcherQueue const& dispatcher,
         Windows::System::DispatcherQueuePriority const priority = Windows::System::DispatcherQueuePriority::Normal) noexcept
@@ -468,5 +469,6 @@
         return resume_foreground(dispatcher);
     }
 #endif
+#endif
 }
 #endif
